php-checks/src/main/java/org/sonar/php/checks/ParameterReassignedToCheck.java

```diff
@@ -0,0 +1,158 @@
+import org.sonar.plugins.php.api.tree.expression.VariableIdentifierTree;
+import org.sonar.plugins.php.api.tree.statement.CatchBlockTree;
+import org.sonar.plugins.php.api.tree.statement.ForEachStatementTree;
+import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;
+
+@Rule(key = "S1226")
+public class ParameterReassignedToCheck extends PHPVisitorCheck {
+  private static final String MESSAGE = "Introduce a new variable instead of reusing the parameter \"%s\".";
+  private static final String SECONDARY_MESSAGE = "Initial value.";
+
+  private final Set<Symbol> variables = new HashSet<>();
+
+  @Override
+  public void visitFunctionDeclaration(FunctionDeclarationTree tree) {
+    Set<Symbol> live = visitFunctionTree(tree);
+    super.visitFunctionDeclaration(tree);
+    clearFixFunctionTree(tree, live);
+  }
+
+  @Override
+  public void visitMethodDeclaration(MethodDeclarationTree tree) {
+    Set<Symbol> live = visitFunctionTree(tree);
+    super.visitMethodDeclaration(tree);
+    clearFixFunctionTree(tree, live);
+  }
+
+  @Override
+  public void visitFunctionExpression(FunctionExpressionTree tree) {
+    Set<Symbol> live = visitFunctionTree(tree);
+    super.visitFunctionExpression(tree);
+    clearFixFunctionTree(tree, live);
+  }
+
+  private Set<Symbol> visitFunctionTree(FunctionTree tree) {
+    ControlFlowGraph cfg = ControlFlowGraph.build(tree, context());
+    if (cfg == null) {
+      return new HashSet<>();
+    }
+
+    LiveVariablesAnalysis analysis = LiveVariablesAnalysis.analyze(cfg, context().symbolTable());
+    Set<Symbol> live = analysis.getLiveVariables(cfg.start()).getIn();
+    for (ParameterTree parameterTree : tree.parameters().parameters()) {
+      if (parameterTree.referenceToken() == null) {
+        Symbol symbol = context().symbolTable().getSymbol(parameterTree.variableIdentifier());
+        if (!live.contains(symbol)) {
+          variables.add(symbol);
+        }
+      }
+    }
+    return live;
+  }
+
+  private void clearFixFunctionTree(FunctionTree tree, Set<Symbol> live) {
+    for (ParameterTree parameterTree : tree.parameters().parameters()) {
+      Symbol symbol = context().symbolTable().getSymbol(parameterTree.variableIdentifier());
+      if (!live.contains(symbol)) {
+        variables.remove(symbol);
+      }
+    }
+  }
+
+  @Override
+  public void visitForEachStatement(ForEachStatementTree tree) {
+    ControlFlowGraph cfg = ControlFlowGraph.build(tree, context());
+    if (cfg == null) {
+      return;
+    }
+
+    Symbol symbol = context().symbolTable().getSymbol(tree.value());
+    LiveVariablesAnalysis analysis = LiveVariablesAnalysis.analyze(cfg, context().symbolTable());
+
+    Set<Symbol> live = analysis.getLiveVariables(cfg.start()).getIn();
+    boolean liveVar = live.contains(symbol);
+
+    if (!liveVar) {
+      variables.add(symbol);
+    }
+
+    super.visitForEachStatement(tree);
+
+    if (!liveVar) {
+      variables.remove(symbol);
+    }
```

Comment:
I'm not sure why we need to modify `variables` here. Removing that only creates 1 more false negative in the unit tests (which I think we could live with ). I think that CFG does not represent `for each` loops very well.

Reply:
The new variable created in the foreach loop must be added to memory so that it is relevant for this issue type in an assignment.